src/serac/physics/utilities/variational_form/tests/tensor_unit_tests.cpp

```diff
@@ -0,0 +1,119 @@
+#include "serac/physics/utilities/variational_form/detail/meta.h"
+#include "serac/physics/utilities/variational_form/tensor.hpp"
+
+using namespace serac;
+
+// note: these tests are actually all static_assert-ions,
+// so if this compiles without error, the tests have already passed
+void basic_tensor_tests()
+{
+  constexpr auto abs = [](auto x) { return (x < 0) ? -x : x; };
+
+  constexpr tensor<double, 3> u = {1, 2, 3};
+  constexpr tensor<double, 4> v = {4, 5, 6, 7};
+
+  constexpr tensor<double, 3, 3> A = make_tensor<3, 3>([](int i, int j) { return i + 2.0 * j; });
+
+  constexpr double sqnormA = 111.0;
+  static_assert(abs(sqnorm(A) - sqnormA) < 1.0e-16);
+
+  constexpr tensor<double, 3, 3> symA = {{{0, 1.5, 3}, {1.5, 3, 4.5}, {3, 4.5, 6}}};
+  static_assert(abs(sqnorm(sym(A) - symA)) < 1.0e-16);
+
+  constexpr tensor<double, 3, 3> devA = {{{-3, 2, 4}, {1, 0, 5}, {2, 4, 3}}};
+  static_assert(abs(sqnorm(dev(A) - devA)) < 1.0e-16);
+
+  constexpr tensor<double, 3, 3> invAp1 = {{{-4, -1, 3}, {-1.5, 0.5, 0.5}, {2, 0, -1}}};
+  static_assert(abs(sqnorm(inv(A + Identity<3>()) - invAp1)) < 1.0e-16);
+
+  constexpr tensor<double, 3> Au = {16, 22, 28};
+  static_assert(abs(sqnorm(dot(A, u) - Au)) < 1.0e-16);
+
+  constexpr tensor<double, 3> uA = {8, 20, 32};
+  static_assert(abs(sqnorm(dot(u, A) - uA)) < 1.0e-16);
+
+  constexpr double uAu = 144;
+  static_assert(abs(dot(u, A, u) - uAu) < 1.0e-16);
+
+  constexpr tensor<double, 3, 4> B = make_tensor<3, 4>([](auto i, auto j) { return 3.0 * i - j; });
+
+  constexpr double uBv = 300;
+  static_assert(abs(dot(u, B, v) - uBv) < 1.0e-16);
+}
```

Comment:
It might be nice to throw in a smoke test for different ways to access elements or rows of the tensor.

Reply:
The accessors are how these functions (`dot`, `det`, `inv`, etc) are implemented, so that is functionality is covered by these tests. Or are you asking for some more examples for users of the different kinds of accessors?